### DISCUSSION TITLE
crash in vips_smartcrop when using attention_x

### DISCUSSION THREAD

**Discussion by dnalor** [libvips/libvips#3587] [General]
## Bug report

I am trying to improve vips_smartcrop's accuracy by using face detection.

code:
```
int res;
if( im->face_detections != NULL && im->num_face_detections > 0 && im->face_detection_area.x2 != 0 && im->face_detection_area.y2 != 0) {
	gint attn_x = im->face_detection_area.x1 + ((im->face_detection_area.x2 - im->face_detection_area.x1) / 2);
	gint attn_y = im->face_detection_area.y1 + ((im->face_detection_area.y2 - im->face_detection_area.y1) / 2);
	PRINT_INFOMSG( "_vips_thumb: smartcrop attn centre: %d %d", attn_x, attn_y);
	res = vips_smartcrop( dest_t, &dest, im->target_width, im->target_height, "interesting", VIPS_INTERESTING_ATTENTION, NULL);
} else {
	res = vips_thumbnail_image( dest_t, &dest, im->target_width, "height", im->target_height, "size", VIPS_SIZE_BOTH, "crop", VIPS_INTERESTING_ATTENTION, NULL);
}
if( res == -1 ) {
	PRINT_ERRMSG( "_vips_thumb: vips_smartcrop/vips_thumbnail_image failed (%s)", vips_error_buffer());
	dest = dest_t;
} else {
	PRINT_INFOMSG( "_vips_thumb: smartcrop ok");
	g_object_unref( dest_t);
}
```
works fine:
```
09:11:38 DEBUG src/vips.c:221: _vips_thumb: smartcrop attn centre: 248 98
09:11:39 DEBUG src/vips.c:233: _vips_thumb: smartcrop ok
```

if I change vips_smartcrop to:
```
res = vips_smartcrop( dest_t, &dest, im->target_width, im->target_height, "interesting", VIPS_INTERESTING_ATTENTION, "attention_x", attn_x, "attention_y", attn_y, NULL);
```
the program crashes right after the output "smartcrop attn centre", no error is printed

Maybe I have misunderstood how attention_x/attention_y is supposed to be used?

**To Reproduce**
happens every time when using vips_smartcrop with attention_x

**Expected behavior**
smartcrop

**Actual behavior**
crash

**Environment**
 - OS: Linux 5.4 (Gentoo), Arm 32bit (armv7a-unknown-linux-gnueabihf)
 - Vips: 8.14.2 (self-compiled with gcc 13)

**Comment by lovell** [comment]
The `attention_x` and `attention_y` arguments are for output and provide the centre around which the crop occurred, so you'd pass in something like `&attn_x` and `&attn_y`.

The docs could probably be a bit clearer about this - PR welcome if you're able.

https://www.libvips.org/API/current/libvips-conversion.html#vips-smartcrop

**Reply by dnalor** [reply]
(sorry for the late reply)
that's a bit disappointing - I had hoped to discourage smartcrop from cutting off parts of faces.
tbh, when reading the docs, it never occurred to me that these are output parms..

anyway, thanks for answering!

**Reply by JunkCoding** [reply]
> The `attention_x` and `attention_y` arguments are for output and provide the centre around which the crop occurred, so you'd pass in something like `&attn_x` and `&attn_y`.
> 
> The docs could probably be a bit clearer about this - PR welcome if you're able.
> 
> https://www.libvips.org/API/current/libvips-conversion.html#vips-smartcrop

As the OP said, that's a shame. It would be nice if you could specify use the X, Y as the spot to focus around.
